src/index.ts

```diff
@@ -44,21 +46,109 @@ interface StorageMetadata extends Record<string, string> {
+	if (!contentType) {
+		throw new HeaderNotDefinedError('"Content-Type" must be defined');
+	}
+
+	if (contentType == MediaType.PRIVATE_TOKEN_REQUEST) {
+		return handleSingleTokenRequest(ctx, request);
+	} else if (contentType == MediaType.ARBITRARY_BATCHED_TOKEN_REQUEST) {
+		// TODO: remember that if content-type is not correct I need to return "Invalid content type", { status: 422 }
+		return handleBatchedTokenRequest(ctx, request);
+	} else {
+		throw new HeaderNotDefinedError(`"Content-Type" must be either "${MediaType.PRIVATE_TOKEN_REQUEST}" or "${MediaType.ARBITRARY_BATCHED_TOKEN_REQUEST}"`);
+	}
+}
+
+export const handleSingleTokenRequest = async (ctx: Context, request: Request) => {
 	ctx.metrics.issuanceRequestTotal.inc({ version: ctx.env.VERSION_METADATA.id ?? RELEASE });
 	const contentType = request.headers.get('content-type');
 	if (!contentType || contentType !== MediaType.PRIVATE_TOKEN_REQUEST) {
 		throw new HeaderNotDefinedError(`"Content-Type" must be "${MediaType.PRIVATE_TOKEN_REQUEST}"`);
 	}
 
 	const buffer = await request.arrayBuffer();
-	const tokenRequest = TokenRequest.deserialize(new Uint8Array(buffer));
+	const tokenRequest = TokenRequest.deserialize(TOKEN_TYPES.BLIND_RSA, new Uint8Array(buffer));
 
 	if (tokenRequest.tokenType !== TOKEN_TYPES.BLIND_RSA.value) {
 		throw new InvalidTokenTypeError();
 	}
 
 	const keyID = tokenRequest.truncatedTokenKeyId.toString();
+	const { sk, pk } = await getKeyPair(ctx, keyID);
+
+	const domain = new URL(request.url).host;
+	const issuer = new Issuer(BlindRSAMode.PSS, domain, sk, pk, { supportsRSARAW: true });
+	const signedToken = await issuer.issue(tokenRequest);
+	ctx.metrics.signedTokenTotal.inc({ key_id: keyID });
+
+	// too verbose with workers observability
+	// once there is a way to filter logpush based on log level, we can consider re-enabling
+	// console.debug(`Token issued successfully for key ${keyID}`);
+
+	return new Response(signedToken.serialize(), {
+		headers: { 'content-type': MediaType.PRIVATE_TOKEN_RESPONSE },
+	});
+};
+
+const handleBatchedTokenRequest = async (ctx: Context, request: Request): Promise<Response> => {
+	try {
+		const buffer = await request.arrayBuffer();
+		const batchedTokenRequest = BatchedTokenRequest.deserialize(new Uint8Array(buffer));
+
+		// Validate that all token requests have the same key ID and correct token type
+		const keyIds = new Set<number>();
+		for (const tokenRequest of batchedTokenRequest.tokenRequests) {
+			if (tokenRequest.tokenType !== TOKEN_TYPES.BLIND_RSA.value) {
+				throw new InvalidTokenTypeError();
+			}
+			keyIds.add(tokenRequest.truncatedTokenKeyId);
+			if (keyIds.size > 1) {
+				throw new Error("Batched token request must have the same key ID");
+			}
+		}
+
+		const firstKey = keyIds.values().next();
+		if (firstKey.done || firstKey.value === undefined) {
+			throw new Error("No valid key ID found in batched token request.");
+		}
+		const keyID = firstKey.value.toString();
```

Comment:
handle the case where there are zero token requests in the batch, and use the first key as a reference directly

I keep keyID as a number as that's closer to the spec. It can be converted to a string where relevant

```suggestion
    if (batchedTokenRequest.length === 0) {
      const responseBytes = (new BatchedTokenResponse([])).serialize()
      return new Response(responseBytes, { headers: { 'Content-Type': MediaType.ARBITRARY_BATCHED_TOKEN_RESPONSE } })
    }
		const keyID = batchedTokenRequest[0].truncatedTokenKeyId;
		for (const tokenRequest of batchedTokenRequest.tokenRequests) {
			if (tokenRequest.tokenType !== TOKEN_TYPES.BLIND_RSA.value) {
				throw new InvalidTokenTypeError();
			}
			if (tokenRequest.truncatedTokenKeyId !== keyID) {
				throw new Error("This issuer only supports BatchedTokenRequest where all requests have the same key ID");
			}
		}
```